pandas/core/groupby/categorical.py

```diff
@@ -72,6 +72,11 @@ def recode_for_groupby(
 
     # sort=False should order groups in as-encountered order (GH-8868)
     cat = c.unique()
+    # exclude nan from indexer for categories
+    take_codes = cat.codes[cat.codes != -1]
+    if cat.ordered:
+        take_codes = np.sort(take_codes)
+    cat = cat.set_categories(cat.categories.take(take_codes))
```

Comment:
Moved from `Categorical.unique`. This keeps groupbys working unchanged.

Reply:
What changes/breaks if you don't include this? 

Normally, the categorical its categories doesn't include any NaNs, so I don't fully understand the comment.

Also, if we don't want to drop unobserved categories in unique, don't we want to do the same change in groupby?

Reply:
The comment refers to nans in `cat` (a `Categorical`), not its categories.

This section is not optimal and is just to keep the same behaviour as previously in groupbys. This can be changed in follow-ups. I don't think this can be untangled without breaking behaviour, e.g. this method of removing unused categories is different than using `remove_unused categories`.

I think we should just accept this smelly bit here here and fix this whole function in followups.

Reply:
But can you explain (or show with an example) what behaviour in groupby would change if this code was not included here?

> This section ... is just to keep the same behaviour as previously in groupbys. This can be changed in follow-ups. I don't think this can be untangled without breaking behaviour

Yes, but we are breaking the behaviour of `unique()` on purpose, so it might be that we want to make the exact same break in groupby?

Reply:
> But can you explain (or show with an example) what behaviour in groupby would change if this code was not included here?

I suppose it is related to the categories of the resulting key index/column after grouping on a categorical:

```
In [45]: df = pd.DataFrame({"key": pd.Categorical(["c", "b", "b"], categories=["a", "b", "c"]), "values": range(3)})

In [46]: df.groupby("key").sum().index
Out[46]: CategoricalIndex(['a', 'b', 'c'], categories=['a', 'b', 'c'], ordered=False, name='key', dtype='category')

In [47]: df.groupby("key", sort=False).sum().index
Out[47]: CategoricalIndex(['c', 'b', 'a'], categories=['c', 'b', 'a'], ordered=False, name='key', dtype='category')

In [51]: df = pd.DataFrame({"key": pd.Categorical(["c", "b", "b"], categories=["a", "b", "c"],  ordered=True), "values": range(3)})

In [52]: df.groupby("key").sum().index
Out[52]: CategoricalIndex(['a', 'b', 'c'], categories=['a', 'b', 'c'], ordered=True, name='key', dtype='category')

In [53]: df.groupby("key", sort=False).sum().index
Out[53]: CategoricalIndex(['b', 'c', 'a'], categories=['b', 'c', 'a'], ordered=True, name='key', dtype='category')
```

So it seems we have a very similar issue here about the order of the resulting categories of the dtype which is not preserved (since it relied on the behaviour of `unique` before)

Reply:
hmm i think @jorisvandenbossche is right here. if you *don't* add the code here, how much to update the groupby tests?